test/e2e/major_upgrade_v3.go

```diff
@@ -97,12 +101,93 @@ func MajorUpgradeToV3(logger *log.Logger) error {
 				testnet.NoError("failed to get header", err)
 				if resp.Header.Version.App == v3.Version {
 					upgradeComplete = true
+					if upgradedHeight == 0 {
+						upgradedHeight = resp.Header.Height
+					}
 				}
 				logger.Printf("height %v", resp.Header.Height)
 				lastHeight = resp.Header.Height
 			}
 		}
 	}
 
+	// check if the timeouts are set correctly
+	rpcNode := testNet.Nodes()[0]
+	client, err := rpcNode.Client()
+	testnet.NoError("failed to get client", err)
+
+	startHeight := upgradeHeightV3 - 5
+	endHeight := upgradedHeight + 5
+
+	type versionDuration struct {
+		dur   time.Duration
+		block *tmtypes.Block
+	}
+
+	blockSummaries := make([]versionDuration, 0, endHeight-startHeight)
+	var prevBlockTime time.Time
+
+	for h := startHeight; h < endHeight; h++ {
+		resp, err := client.Block(ctx, &h)
+		testnet.NoError("failed to get header", err)
+		blockTime := resp.Block.Time
+
+		if h == startHeight {
+			if resp.Block.Version.App != v2.Version {
+				return fmt.Errorf("expected start height %v was app version 2", startHeight)
+			}
+			prevBlockTime = blockTime
+			continue
+		}
+
+		blockDur := blockTime.Sub(prevBlockTime)
+		prevBlockTime = blockTime
+		blockSummaries = append(blockSummaries, versionDuration{dur: blockDur, block: resp.Block})
+	}
+
+	preciseUpgradeHeight := 0
+	multipleRounds := 0
+	for _, b := range blockSummaries {
+
+		// check for the precise upgrade height and skip, as the block time
+		// won't match due to the off by 1 nature of the block time.
+		if b.block.Version.App == v3.Version && preciseUpgradeHeight == 0 {
+			preciseUpgradeHeight = int(b.block.Height)
+			continue
+		}
+
+		// don't test heights with multiple rounds as the times are off and fail
+		// later if there are too many
+		if b.block.LastCommit.Round > 0 {
+			multipleRounds++
+			continue
+		}
+
+		if b.dur > appconsts.GetTimeoutCommit(b.block.Version.App) {
+			return fmt.Errorf(
+				"block was too slow for corresponding version: version %v duration %v upgrade height %v height %v",
+				b.block.Version.App,
+				b.dur,
+				preciseUpgradeHeight,
+				b.block.Height,
+			)
+		}
+
+		if b.dur < appconsts.GetTimeoutCommit(b.block.Version.App) {
+			return fmt.Errorf(
+				"block was too fast for corresponding version: version %v duration %v upgrade height %v height %v",
+				b.block.Version.App,
+				b.dur,
+				preciseUpgradeHeight,
+				b.block.Height,
+			)
+		}
```

Comment:
[question] it looks like this will throw an error if b.dur doesn't exactly match the timeout commit which seems extremely fragile. I would expect this test to pass if blocks times are ~6 seconds which is > the timeout commit of 3.5 seconds. Maybe I'm missing something.

Reply:
ahh whoops yeah good point, will fix in a flup

knuu isn't working still so this test was only manually performed and this wasn't hit

it needs to be 
```go
if b.dur > appconsts.GetTimeoutCommit([the other app version]) {
}
```

Reply:
https://github.com/celestiaorg/celestia-app/pull/3882/commits/9669f8c996b445637132231bff70ed1d53c0f6ab